libs/execution_context/ebpf_maps.c

```diff
@@ -985,25 +921,64 @@ _get_key_state(_In_ const ebpf_core_lru_map_t* map, _In_ const ebpf_lru_entry_t*
     }
 }
 
+/**
+ * @brief Helper function to merge the hot list into the cold list if the hot list size exceeds the hot list limit.
+ * Resets the hot list size and increments the current generation.
+ *
+ * @param[in,out] map Pointer to the map.
+ */
+_Requires_lock_held_(map->lock) static void _merge_hot_into_cold_list_if_needed(_Inout_ ebpf_core_lru_map_t* map)
+{
+    if (map->hot_list_size <= map->hot_list_limit) {
+        return;
+    }
+
+    ebpf_list_entry_t* list_entry = map->hot_list.Flink;
+    ebpf_list_remove_entry(&map->hot_list);
+    ebpf_list_append_tail_list(&map->cold_list, list_entry);
+
+    ebpf_list_initialize(&map->hot_list);
+    map->hot_list_size = 0;
+    map->current_generation++;
+}
+
 /**
  * @brief Helper function to insert an entry into the hot list if it is in the cold list and update the hot list size.
  *
  * @param[in,out] map Pointer to the map.
  * @param[in,out] entry Entry to insert into the hot list.
  */
-_Requires_lock_held_(map->lock) static void _insert_into_hot_list(
-    _Inout_ ebpf_core_lru_map_t* map, _Inout_ ebpf_lru_entry_t* entry)
+static void
+_insert_into_hot_list(_Inout_ ebpf_core_lru_map_t* map, _Inout_ ebpf_lru_entry_t* entry)
 {
+    bool lock_held = false;
     ebpf_lru_key_state_t key_state = _get_key_state(map, entry);
+    ebpf_assert(key_state == EBPF_LRU_KEY_HOT || key_state == EBPF_LRU_KEY_COLD || key_state == EBPF_LRU_KEY_DELETED);
+
     // Skip if not in the cold list.
     // If not yet initialized, it will be added to the hot list when initialized.
     // If already deleted, don't add it to the hot list.
     if (key_state != EBPF_LRU_KEY_COLD) {
-        return;
+        goto Exit;
```

Comment:
Why are we're checking ```key_state``` here without holding the map lock and yet do the same check immediately after taking that lock? Curious why we couldn't get the key state under the lock to begin with?

Reply:
We check it outside the lock to avoid having to take the lock if the item is already in the hot list. The value could be stale, but there are two possible cases, based on what the actual state is:
1) Item is actually cold but is seen as not cold -> LRU update for this item is skipped on this update.
2) Item is not cold but is seen as cold -> Lock is taken, item is rechecked, and function exits.